go/vt/vtgate/endtoend/vstream_test.go

```diff
@@ -258,6 +258,129 @@ func TestVStreamCopyBasic(t *testing.T) {
+
+	expectedKs1EventNum := 2 /* num shards */ * (9 /* begin/field/vgtid:pos/4 rowevents avg/vgitd: lastpk/commit) */ + 3 /* begin/vgtid/commit for completed table */ + 1 /* copy operation completed */)
+	expectedKs2EventNum := 2 /* num shards */ * (6 /* begin/field/vgtid:pos/1 rowevents avg/vgitd: lastpk/commit) */ + 3 /* begin/vgtid/commit for completed table */ + 1 /* copy operation completed */)
+	expectedFullyCopyCompletedNum := 1
+
+	cases := []struct {
+		name                    string
+		shardGtid               *binlogdatapb.ShardGtid
+		expectedEventNum        int
+		expectedCompletedEvents []string
+	}{
+		{
+			name: "copy from all keyspaces",
+			shardGtid: &binlogdatapb.ShardGtid{
+				Keyspace: "/.*",
+			},
+			expectedEventNum: expectedKs1EventNum + expectedKs2EventNum + expectedFullyCopyCompletedNum,
+			expectedCompletedEvents: []string{
+				`type:COPY_COMPLETED keyspace:"ks" shard:"-80"`,
+				`type:COPY_COMPLETED keyspace:"ks" shard:"80-"`,
+				`type:COPY_COMPLETED keyspace:"ks2" shard:"-80"`,
+				`type:COPY_COMPLETED keyspace:"ks2" shard:"80-"`,
+				`type:COPY_COMPLETED`,
+			},
+		},
+		{
+			name: "copy from all shards in the keyspace",
+			shardGtid: &binlogdatapb.ShardGtid{
+				Keyspace: "ks",
+			},
+			expectedEventNum: expectedKs1EventNum + expectedFullyCopyCompletedNum,
+			expectedCompletedEvents: []string{
+				`type:COPY_COMPLETED keyspace:"ks" shard:"-80"`,
+				`type:COPY_COMPLETED keyspace:"ks" shard:"80-"`,
+				`type:COPY_COMPLETED`,
+			},
+		},
+	}
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			gconn, conn, mconn, closeConnections := initialize(ctx, t)
+			defer closeConnections()
+
+			var vgtid = &binlogdatapb.VGtid{}
+			vgtid.ShardGtids = []*binlogdatapb.ShardGtid{c.shardGtid}
+			reader, err := gconn.VStream(ctx, topodatapb.TabletType_PRIMARY, vgtid, filter, flags)
+			_, _ = conn, mconn
+			if err != nil {
+				t.Fatal(err)
+			}
+			require.NotNil(t, reader)
+			var evs []*binlogdatapb.VEvent
+			var completedEvs []*binlogdatapb.VEvent
+			for {
+				e, err := reader.Recv()
+				switch err {
+				case nil:
+					evs = append(evs, e...)
+
+					for _, ev := range e {
+						if ev.Type == binlogdatapb.VEventType_COPY_COMPLETED {
+							completedEvs = append(completedEvs, ev)
+						}
+					}
+
+					printEvents(evs) // for debugging ci failures
+
+					if len(evs) == c.expectedEventNum {
+						// The arrival order of COPY_COMPLETED events with keyspace/shard is not constant.
+						// On the other hand, the last event should always be a fully COPY_COMPLETED event.
+						// That's why the sort.Slice doesn't have to handle the last element in completedEvs.
+						sort.Slice(completedEvs[:len(completedEvs)-1], func(i, j int) bool {
+							if completedEvs[i].GetKeyspace() != completedEvs[j].GetKeyspace() {
+								return completedEvs[i].GetKeyspace() < completedEvs[j].GetKeyspace()
+							}
+							return completedEvs[i].GetShard() < completedEvs[j].GetShard()
+						})
```

Comment:
It makes sense. This is not identical to the other one, but it can be used for both cases, so I created a utility function by [a013980](https://github.com/vitessio/vitess/pull/11909/commits/a013980946e119147d1a4b68af1f6ca045b1aaa9).